### ISSUE TITLE
Latest release erroring on sphinx-build with: "the sphinxcontrib.spelling extension is not safe for parallel reading"

### ISSUE THREAD

**Issue by alex** [sphinx-contrib/spelling#55]
```
Traceback (most recent call last):
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/application.py", line 348, in build
    self.builder.build_update()
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 297, in build_update
    self.build(to_build,
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/builders/__init__.py", line 311, in build
    updated_docnames = set(self.read())
  File "/opt/python/3.8.1/lib/python3.8/contextlib.py", line 120, in __exit__
    next(self.gen)
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/util/logging.py", line 213, in pending_warnings
    memhandler.flushTo(logger)
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/util/logging.py", line 178, in flushTo
    logger.handle(record)
  File "/opt/python/3.8.1/lib/python3.8/logging/__init__.py", line 1587, in handle
    self.callHandlers(record)
  File "/opt/python/3.8.1/lib/python3.8/logging/__init__.py", line 1649, in callHandlers
    hdlr.handle(record)
  File "/opt/python/3.8.1/lib/python3.8/logging/__init__.py", line 946, in handle
    rv = self.filter(record)
  File "/opt/python/3.8.1/lib/python3.8/logging/__init__.py", line 807, in filter
    result = f.filter(record)
  File "/home/travis/build/pyca/cryptography/.tox/docs/lib/python3.8/site-packages/sphinx/util/logging.py", line 417, in filter
    raise SphinxWarning(message)
sphinx.errors.SphinxWarning: the sphinxcontrib.spelling extension is not safe for parallel reading
```

This is happening on a build that's not even using the sphinx builder:

https://travis-ci.org/github/pyca/cryptography/jobs/701818891#L454

**Comment by dhellmann** [comment]
This extension doesn't work with parallel builds because it attaches data to pages during the read phase and if different processes perform the read vs. the build then the data isn't available. Until that's resolved somehow we shouldn't advertise parallel build support.

I'm not sure why the setting for this extension would be affecting a run that isn't using the spelling builder, but I don't think that's a bug in this extension.

Is there some way to only enable the extension when it's used (that seems like it presents a chicken-and-egg problem)?

Is there a way to ignore the parallel build setting for the extension unless the builder is being used?

**Comment by dhellmann** [comment]
See #34 for more history.

**Comment by alex** [comment]
FWIW, this is definitely caused by something that changed in 5.1.2, if we disallow that version the parallel HTML build works fine. I guess something must have changed in the initialization process or something like that?

**Comment by jdillard** [comment]
@alex you are running sphinx-build with the `-W` flag, so the warning about this extension not being not safe for parallel reading is turned into an error, and thus a failure. But I also believe your point about it running on builders other than `spelling` is a bug with this extension, I was actually looking into that for a different issue I was having. So, in the end, I believe there should be a way for this extension to not cause your `html` builder to throw a warning and thus fail.

@dhellmann you can see this spelling extension initialize on the `html` builder in the linked Travis-CI log:

```
docs run-test: commands[0] | sphinx-build -j4 -T -W -b html -d /home/travis/build/pyca/cryptography/.tox/docs/tmp/doctrees docs docs/_build/html
Running Sphinx v3.0.4
Initializing Spelling Checker
```
I was going to try to implement a builder allow list, similar to how the `sphinx-tabs` extension does it, but have been unsuccessful: https://github.com/djungelorm/sphinx-tabs/blob/6766b54ae7f3e928ccfa31a4cd7cfaf09e96400b/sphinx_tabs/tabs.py#L31

**Comment by dhellmann** [comment]
> FWIW, this is definitely caused by something that changed in 5.1.2, if we disallow that version the parallel HTML build works fine. I guess something must have changed in the initialization process or something like that?

Yes, in 5.1.2 we changed the flag in this extension that claims it supports parallel builds because it doesn't. So it's definitely exposed in your CI job with that release. That's not a bug in this project, though, is it?

**Comment by alex** [comment]
Oh, I'd assumed that the change was something around initialization, not around the parallel support flag.

In that case I'm honestly not sure whether this is a spelling or a sphinx-itself bug.

**Comment by dhellmann** [comment]
> @alex you are running sphinx-build with the `-W` flag, so the warning about this extension not being not safe for parallel reading is turned into an error, and thus a failure. But I also believe your point about it running on builders other than `spelling` is a bug with this extension, I was actually looking into that for a different issue I was having, but wasn't quite sure yet. So, in the end, I believe there should be a way for this extension to not cause your `html` builder to throw a warning and thus fail.
> 
> @dhellmann you can see this spelling extension initialize on the `html` builder in the linked Travis-CI log:
> 
> ```
> docs run-test: commands[0] | sphinx-build -j4 -T -W -b html -d /home/travis/build/pyca/cryptography/.tox/docs/tmp/doctrees docs docs/_build/html
> Running Sphinx v3.0.4
> Initializing Spelling Checker
> ```
> 
> I was going to try to implement a compatible builder list, similar to how the `sphinx-tabs` extension does it, but have been unsuccessful: https://github.com/djungelorm/sphinx-tabs/blob/6766b54ae7f3e928ccfa31a4cd7cfaf09e96400b/sphinx_tabs/tabs.py#L31

The setup() function is invoked when the extension is imported by Sphinx. If there's a way to initialize the extension only when it's actually used, I would definitely switch to that, but I'm not aware of a different hook that can be used. The builder-inited event happens after the builder is initialized, and this extension is an alternative builder so that won't work. The config-inited is similarly after configuration is done, which means we can't add the config options used by the builder. Everything else happens even later in the process.

**Comment by dhellmann** [comment]
> Oh, I'd assumed that the change was something around initialization, not around the parallel support flag.
> 
> In that case I'm honestly not sure whether this is a spelling or a sphinx-itself bug.

I'm sure it's unusual to have a builder that doesn't support parallel reads. I would love to fix that, but I'm not sure how. I haven't had a chance to look into how other builders attach data to nodes in a way that persists across processes, but I think that's really the problem.

**Comment by jdillard** [comment]
> The setup() function is invoked when the extension is imported by Sphinx. If there's a way to initialize the extension only when it's actually used, I would definitely switch to that, but I'm not aware of a different hook that can be used. The builder-inited event happens after the builder is initialized, and this extension is an alternative builder so that won't work. The config-inited is similarly after configuration is done, which means we can't add the config options used by the builder. Everything else happens even later in the process.

Ah, great points. It seems adding the extension at runtime might be best for CI/CD pipelines, and while I'm not sure which one has the least technical debt, overriding the `extensions` config value with the `-D` flag is one example:

```
sphinx-build -b spelling -D extensions=sphinx_extension1,sphinx_extension2,sphinxcontrib.spelling [...]
```

**Comment by alex** [comment]
FWIW, we're just working around this by disabling parallel builds, since docs are far from our slowest CI job anyways.

**Comment by dhellmann** [comment]
I would be happy to accept doc updates to describe any of the work-arounds you develop. As well as, of course, code changes to make parallel builds work. :-) 

Maybe those doc notes could go early in `docs/source/customize.rst` to make them easy to spot when someone is getting started.

**Comment by webknjaz** [comment]
> ```
> sphinx-build -b spelling -D extensions=sphinx_extension1,sphinx_extension2,sphinxcontrib.spelling [...]
> ```

FWIW it's also possible to make it conditional in `conf.py`: https://github.com/cherrypy/cheroot/commit/63097fbfd40eda733a14893a6ed7d1e370a8cde1.

**Comment by dhellmann** [comment]
@webknjaz that's clever, thanks for sharing. I will work on adding something to the documentation based on that.

**Comment by dhellmann** [comment]
@alex @jdillard @webknjaz I think I've worked out how to actually support parallel builds. Could you try the code in #59 to see if it works for you before I merge it and produce a release?

**Comment by webknjaz** [comment]
I can confirm that #59 works and also fixes #34.
